Submit diffs instead of full phrase manifests

camertron opened the pull request:
This PR refactors the puller, pusher, and completer to submit diffs instead of full phrase manifests. If only a few phrases change in a commit, it doesn’t really make sense to upload every single phrase. We’re hoping this will reduce noise in Smartling and not overwhelm our translators too much. It also means that phrases that haven’t changed but have been “excluded” in Smartling will not be re-uploaded and appear in the Smartling authorization view.

The puller needs a bit more explanation. The Rosette database still requires a full phrase manifest for each commit that touches tracked files (eg. en.yml). To meet that need but still only push diffs, the puller needs to serialize a full phrase manifest, upload it to smartling, download each language, then delete the file. In other words, each pull file is temporary and only used to download phrases. No usernames or branch names are attached to these uploads, and they should only appear briefly in Smartling’s authorize view. To further mitigate confusion, Matt and I have given some thought to creating a separate Smartling project that shares a translation memory with the original project. Using this auxiliary project would give us the ability to upload phrases and download translations as well as avoid cluttering up the files and authorization views in the Smartling interface.

camertron committed:
Refactor all the things

Instead of pushing full phrase lists, push diffs instead. Pull by
uploading full phrase lists, downloading translations for each locale,
and deleting the file. Refactor the completer a bit.

camertron committed:
Add specs

jdoconnor commented on lib/rosette/integrations/smartling-integration/smartling_completer.rb:
```diff
@@ -1,49 +1,164 @@
+
+        def initialize(rosette_config)
+          @rosette_config = rosette_config
+          @thread_pool_size = DEFAULT_THREAD_POOL_SIZE
```
this isn't part of the config?

camertron replied:
No, the thread pool size is determined by the caller.

jdoconnor commented on lib/rosette/integrations/smartling-integration/smartling_completer.rb:
```diff
@@ -1,49 +1,164 @@
+          build_completion_maps.each do |file_uri, locales|
+            if locales.all? { |_, v| v.complete? }
+              begin
+                file = locales[repo_config.locales.first.code]
```
nitpick: without validation, the repo config could not have any locales and cause an error here.

camertron replied:
Right, I almost put a guard in here, but didn't think it was really necessary considering the config should always have locales. If someone forgets to add them however, the error message is going to be confusing. I'll put something in to check for them and raise a `NoConfiguredLocalesError` or something.

jdoconnor commented on lib/rosette/integrations/smartling-integration/smartling_completer.rb:
```diff
@@ -1,49 +1,164 @@
+
+        def build_completion_maps
+          if thread_pool_size > 1
+            build_completion_maps_asynchronously
```
should the asynchronously work for only 1 thread? Maybe an opportunity to simplify and only have one build_completion_maps

camertron replied:
That's how we originally had it, but it made debugging tests more difficult. Trying to insert breakpoints and then having two threads (main and spawned) writing to the console was confusing. We could instead change this to `if thread_pool_size > 0` then build asynchronously, does that make more sense?

jdoconnor replied:
It's fine either way, but I saw a place that may make sense to redactor for
ease of reading
On Feb 11, 2015 8:12 AM, "Cameron Dutro" notifications@github.com wrote:

> In lib/rosette/integrations/smartling-integration/smartling_completer.rb
> https://github.com/rosette-proj/rosette-integration-smartling/pull/1#discussion_r24507061
> :
> 
> > -              smartling_apis[repo_name].delete(file_uri)
> > -        def complete_log(file)
> > -          rosette_config.datastore.add_or_update_commit_log(
> > -            file.repo_name, file.commit_id, nil, Rosette::DataStores::PhraseStatus::TRANSLATED
> > -          )
> > -        end
> >   +
> > -        def build_completion_maps
> > -          if thread_pool_size > 1
> > -            build_completion_maps_asynchronously
> 
> That's how we originally had it, but it made debugging tests more
> difficult. Trying to insert breakpoints and then having two threads (main
> and spawned) writing to the console was confusing. We could instead change
> this to if thread_pool_size > 0 then build asynchronously, does that make
> more sense?
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/rosette-proj/rosette-integration-smartling/pull/1/files#r24507061
> .

jdoconnor commented on lib/rosette/integrations/smartling-integration/smartling_completer.rb:
```diff
@@ -1,49 +1,164 @@
+            current_completed_count = pool.completed_task_count
+
+            if current_completed_count > last_completed_count
+              puts "#{current_completed_count} of #{total} completed"
```
you may be able to achieve not caring about how complete you are (making simpler code) by doing a 

```
puts total
```

and in the merge_mutex maintain a counter of merged maps

less elegant output, but less code.

jdoconnor replied:
do we want puts or to use the logger?

camertron replied:
We do want to use the logger, good catch :)

Keeping track of the merged maps is a cool idea, I'll see if I can refactor things a bit.

jdoconnor commented on lib/rosette/integrations/smartling-integration/smartling_puller.rb:
```diff
@@ -1,135 +1,217 @@
       class SmartlingPuller
 
-        attr_reader :rosette_config, :smartling_apis, :completed_files_map, :locale
+        DEFAULT_THREAD_POOL_SIZE = 15
```
this thread pool size is different than above. coincidence? on purpose because they're different operations?

camertron replied:
Ya they're different operations. I think we cap our MySQL connections to 16, so I chose 15 to leave some wiggle room. I think however it might be better to use 10 here too to leave more than just a single free connection. What do you think?

jdoconnor replied:
It seems odd to have two different settings unless they are VERY different,
but thats just me thinking as a consumer of it
On Feb 11, 2015 8:21 AM, "Cameron Dutro" notifications@github.com wrote:

> In lib/rosette/integrations/smartling-integration/smartling_puller.rb
> https://github.com/rosette-proj/rosette-integration-smartling/pull/1#discussion_r24507955
> :
> 
> >  module Rosette
> >    module Integrations
> >      class SmartlingIntegration < Integration
> >        class SmartlingPuller
> > -        attr_reader :rosette_config, :smartling_apis, :completed_files_map, :locale
> > -        DEFAULT_THREAD_POOL_SIZE = 15
> 
> Ya they're different operations. I think we cap our MySQL connections to
> 16, so I chose 15 to leave some wiggle room. I think however it might be
> better to use 10 here too to leave more than just a single free connection.
> What do you think?
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/rosette-proj/rosette-integration-smartling/pull/1/files#r24507955
> .

camertron replied:
Right that makes sense. Will change back to 10.

jdoconnor commented on lib/rosette/integrations/smartling-integration/smartling_completer.rb:
```diff
@@ -1,49 +1,164 @@
+        DEFAULT_THREAD_POOL_SIZE = 10
+
+        attr_reader :rosette_config, :repo_config
+        attr_reader :thread_pool_size
```
why are these readers when we have a set_thread_pool_size and set_repo_config?

camertron replied:
Are you thinking they should be accessors instead? The `set_` methods enable chaining because they return `self`, and the readers allow reading the values later.

jdoconnor commented on lib/rosette/integrations/smartling-integration/smartling_puller.rb:
```diff
@@ -1,135 +1,217 @@
+        end
 
-            potential_files = rosette_config.datastore.file_list_for_repo(repo_name)
+        def set_serializer_id(serializer_id)
```
same as above... set_<variable> and having a reader for that seems odd to me (in ruby, that is)

camertron replied:
Yeah it's all about the chaining. All the `Command` classes in `Rosette::Core` follow the same pattern. I guess maybe I'm getting too Java up in here (this is pretty much how Java factories work too).

jdoconnor commented on lib/rosette/integrations/smartling-integration/smartling_puller.rb:
```diff
@@ -1,135 +1,217 @@
+          head_refs = repo_config.repo.all_head_refs
+          head_refs.each_with_index do |head_ref, idx|
+            pull_head(head_ref)
+            logger.info(
```
consider logging in json, its the new hotness (plus it is easy to group and search logs)

camertron replied:
Oh yeah that's a cool idea. We don't currently log in JSON anywhere in the project, but it wouldn't be that hard to switch over. What kind of format would you suggest? Maybe something like `{"message":"I broke!"}`?

jdoconnor replied:
take a look at https://github.com/lumoslabs/clerk/pull/110

the gem basically does this 
logger.info "foo" #=> { "message": "foo" ... }
logger.info { i_like: :cake } #=> { "i_like": "cake" }

jdoconnor commented on lib/rosette/integrations/smartling-integration/smartling_pusher.rb:
```diff
@@ -4,101 +4,64 @@ module Rosette
-              file_type_for(serializer_id), {
-                approved: smartling_api.preapprove_translations?
-              }
+          if phrases.size > 0
```
consider using a guard

camertron replied:
You mean like `return unless phrases.size > 0`?

jdoconnor commented on lib/rosette/integrations/smartling-integration/smartling_pusher.rb:
```diff
@@ -4,101 +4,64 @@ module Rosette
-          Rosette.logger.error('Caught an exception while pushing to Smartling API.')
-          Rosette.logger.error("#{ex.message}\n#{ex.backtrace.join("\n")}")
-          raise ex
+          rosette_config.error_reporter.report_error(ex)
```
you may be able to use a common method to get at the logger across the various classes.

camertron replied:
The logger is different from the error reporter. The logger just prints to the console (or to a file), but the error reporter connects to Rollbar. The error reporter is configured at the `Rosette::Configurator` level, not at the `Rosette` level.

jdoconnor commented on lib/rosette/integrations/smartling-integration/smartling_uploader.rb:
```diff
@@ -0,0 +1,116 @@
+        end
+
+        def integration_config
+          @integration_config ||=
```
why two lines?

camertron replied:
Thought they were too long, looks silly now :)

camertron committed:
Pass and use logger in SmartlingCompleter

Commit:
Reflect new SmartlingPusher signature

Commit:
Get repo config from repo name

Commit:
Set PENDING state even when there are no phrases

Happens when there are actually no phrases to push up to Smartling, but
we still need to update the status

Commit:
Pull all PENDING commits instead of just head refs

Commit:
Update translation counts in the completer

Commit:
Fix specs

Commit:
Delete unused private method

Commit:
Run asynchronously if thread pool > 0

Commit:
Put methods on one line

Commit:
Add more logging to completer

Commit:
Don't update the commit log in completer

camertron committed:
Add SmartlingTmpFile

camertron committed:
Update commit_log in puller

camertron committed:
Add exponential back off to retrier

camertron committed:
Make puller download with back off

camertron committed:
Fix specs

camertron committed:
Add back off to SmartlingPuller#file_status_for

camertron committed:
Add more retries

camertron committed:
Increase retries a bit again

camertron committed:
Refactor pusher to process all untranslated commits

camertron committed:
Remove after commit hook :(

camertron committed:
Use refactored datastore method in puller

To iterate over all pending commits

camertron committed:
Add attr_reader for logger

camertron committed:
Fix status method calls in puller

camertron committed:
Send more information when logging errors
